apollo-federation-types/src/composition/mod.rs

```diff
@@ -88,11 +89,35 @@ impl From<BuildHint> for Issue {
     }
 }
 
+/// Rover and GraphOS expect messages to start with `[subgraph name]`. (They
+/// don't actually look at the `locations` field, sadly). This will prepend
+/// the subgraph name if there's exactly one. If there's more than one, it's
+/// probably a composition issue that's not attributable to a single subgraph,
+/// and GraphOS will show "[subgraph unknown]", which is also not correct.
+fn maybe_prepend_subgraph(message: &str, locations: &[SubgraphLocation]) -> String {
+    if message.starts_with('[') {
+        return message.to_string();
+    }
+    let unique_subgraphs = locations
+        .iter()
+        .filter_map(|l| l.subgraph.clone())
```

Comment:
Nitpick, but you can avoid cloning every item:
```suggestion
        .filter_map(|l| l.subgraph.as_ref())
```